### PR TITLE
[FEATURE] Add CLI Command

### PR DESCRIPTION
**Pull request by jdoubleu**
Adds a command for use with console to precompile and profile fluid templates.

See also https://github.com/NamelessCoder/typo3-cms-fluid-precompiler/issues/2

### CODE REVIEW

File: Classes/Command/FluidPrecompileCommandController.php [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+     * @throws \Exception
+     */
+    public function compileCommand($extension = null, $fail = true, $onlyFailed = true, $limit = null)
+    {
+        $lastError = null;
+
+        try {
+            $result = $this->getPrecompilerService()->warmup($extension);
+        } catch (\Exception $e) {
+            $result = [];
```
Comment (NamelessCoder) [inline Classes/Command/FluidPrecompileCommandController.php]: This looks like a misunderstanding about what `$fail` actually does.

It does not suppress errors. It makes the warmup fail if any one template could not compile - which when it happens, does not raise an exception. So no try-catch here, but instead, a `throw` down below if there is one or more than failed templates.

**Review by NamelessCoder — CHANGES_REQUESTED** [review]
Everything else looks pretty damn good, the one thing that could improve it is to use strict mode (up to 7.0 so no nullable). Very minor since only a single return type annotation could be added anyway.

The other comment in code is however more important to address!

File: Classes/Command/FluidPrecompileCommandController.php [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Extbase\Mvc\Controller\CommandController;
+use TYPO3\CMS\Extbase\Object\ObjectManager;
+use TYPO3Fluid\Fluid\Core\Cache\FluidCacheWarmupResult;
+use TYPO3Fluid\Fluid\Core\Rendering\RenderingContextInterface;
+
+/**
+ * Class FluidPrecompileCommandController
+ */
+class FluidPrecompileCommandController extends CommandController
```
Comment (mbrodala) [inline Classes/Command/FluidPrecompileCommandController.php]: I suggest making this a Symfony command instead, this way we skip the whole Extbase overhead. The command can be used only with TYPO3v8 or newer anyways.

**Review by mbrodala — CHANGES_REQUESTED** [review]

File: Classes/Command/FluidPrecompileCommandController.php (hunk 2 above)
Comment (NamelessCoder) [reply Classes/Command/FluidPrecompileCommandController.php]: Good point, let's do that.

Comment (jdoubleu) [reply Classes/Command/FluidPrecompileCommandController.php]: I'd normally agree with you on this.
I prefer a Symfony Command, too, because TYPO3's Command API is very limited.

However I wanted to stay compatible with the TYPO3 Core (or Extbase) and especially the [TYPO3 Console](https://github.com/TYPO3-Console/TYPO3-Console).

Another benefit of the Command Controller is, that it can be used by the scheduler extension.

A Symfony Command would also require some sort of Symfony Application bootstrapping (even if its a oneliner). How would you've done this? Creating an extra _bin_ script?

I'd suggest to add an additional Symfony Command but also keep the TYPO3 Command.

File: Classes/Command/FluidPrecompileCommandController.php (hunk 1 above)
Comment (jdoubleu) [reply Classes/Command/FluidPrecompileCommandController.php]: Sorry, I can't quite follow. Exceptions will be thrown at the end of the function if `--fail` was provided.

I understand the `--fail` flag as following:
If there's any kind of an error (e.g. compiling, TYPO3 or "uncompilable templates") **and** the `--fail` flag was provided the program should exit with a non-zero return code. In this case a thrown exception would cause this.
But when the `--fail` flag isn't provided it shouldn't, because the program was told not to fail.

Or would you like to see the application "failing" if an internal error (i.e. the compiling failed) occurs while staying silent for all contextual error (i.e. "uncompilable templates") when no `--fail` flag was provided.

**Comment by jdoubleu** [comment]
Additional I'd like to add tests. This extension has no tests yet, so I need to setup/configure the testing environment/framework.

File: Classes/Command/FluidPrecompileCommandController.php (hunk 2 above)
Comment (mbrodala) [reply Classes/Command/FluidPrecompileCommandController.php]: TYPO3 [natively supports](https://docs.typo3.org/typo3cms/extensions/core/8.7/singlehtml/Index.html#feature-73042-introduce-native-support-for-symfony-console) and uses Symfony commands. And starting from TYPO3v9 you have a [scheduler task for executing registered Symfony tasks](https://docs.typo3.org/typo3cms/extensions/core/latest/singlehtml/#feature-79462-introduce-scheduler-task-to-execute-console-command) too.

TL;DR: it's easier than you think. ;-)

Comment (jdoubleu) [reply Classes/Command/FluidPrecompileCommandController.php]: Okay, I didn't know. Sadly, the core documentation on this very poor and as far as I've checked, the extbase commands rely on the TYPO3 Command.

I'm going to migrate the CommandController to a Symfony Command.

**Comment by jdoubleu** [comment]
I've migrated the CommandController into native Symfony Commands.

One downside is, that these commands are no longer available in the [TYPO3 Console](https://github.com/TYPO3-Console/TYPO3-Console).

One thing I noticed was that this extension already defines some TYPO3 core commands like `backend:lock` (see [`Commands.php`](https://github.com/NamelessCoder/typo3-cms-fluid-precompiler-module/blob/02ec114d94051cb4d1aa9ec88d24385cd85a2ab9/Configuration/Commands.php)) causing an exception for the `typo3` command in TYPO3 CMS projects because this Command was already registered:
```
# vendor/bin/typo3
TYPO3\CMS\Core\Console\CommandNameAlreadyInUseException: Command "backend:lock" registered by "cms_fluid_precompiler_module" is already in use in file vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandRequestHandler.php on line 148
Stack trace:
  1. TYPO3\CMS\Core\Console\CommandNameAlreadyInUseException->() vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandRequestHandler.php:148
  2. TYPO3\CMS\Core\Console\CommandRequestHandler->populateAvailableCommands() vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandRequestHandler.php:72
  3. TYPO3\CMS\Core\Console\CommandRequestHandler->handleRequest() vendor/typo3/cms/typo3/sysext/core/Classes/Core/Bootstrap.php:315
  4. TYPO3\CMS\Core\Core\Bootstrap->handleRequest() vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandApplication.php:75
  5. TYPO3\CMS\Core\Console\CommandApplication->run() vendor/typo3/cms-cli/typo3:22
  6. {closure}() vendor/typo3/cms-cli/typo3:23
```

File: Classes/Command/FluidPrecompileCommandController.php (hunk 1 above)
Comment (jdoubleu) [reply Classes/Command/FluidPrecompileCommandController.php]: I've reworked the `--fail` logic:
Exceptions thrown while compiling/warmup won't be catched anymore.
Only if `--faill` was provided **and** there were uncompilable templates, a custom exception will be thrown which causes a non-zero status code.

See https://github.com/NamelessCoder/typo3-cms-fluid-precompiler-module/pull/6/commits/7215c5d13366cdd2d56357e7cb58a0732ec595a1

**Comment by mbrodala** [comment]
@jdoubleu Symfony commands registered for TYPO3 work perfectly fine in the TYPO3 Console, at least in 4.x and 5.x.

And I don't know what the current `Commands.php` is supposed to do. As you can see what it currently does is invalid. @NamelessCoder can you shed some light on this?

**Comment by jdoubleu** [comment]
> @jdoubleu Symfony commands registered for TYPO3 work perfectly fine in the TYPO3 Console, at least in 4.x and 5.x.

Okay, even better. Didn't really test it with the TYPO3 Console. Maybe this was just a mistake on my side.

@mbrodala @NamelessCoder could you please review the new changes?

File: Classes/Command/FluidPrecompilerServiceTrait.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * Receives an instance of the FluidPrecompilerService
+     *
+     * @return FluidPrecompilerService
+     */
+    protected function getPrecompilerService()
+    {
+        $objectManager = GeneralUtility::makeInstance(ObjectManager::class);
+
+        return $objectManager->get(FluidPrecompilerService::class);
```
Comment (mbrodala) [inline Classes/Command/FluidPrecompilerServiceTrait.php]: There's no need for Extbase here, you can just use `GeneralUtility::makeInstance(FluidPrecompilerService::class)` straight here.

File: Classes/Command/FluidPrecompilerServiceTrait.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+namespace NamelessCoder\CmsFluidPrecompilerModule\Command;
+
+use NamelessCoder\CmsFluidPrecompilerModule\Service\FluidPrecompilerService;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Extbase\Object\ObjectManager;
+
+/**
+ * Trait used by commands to get the FluidPrecompilerService
+ */
+trait FluidPrecompilerServiceTrait
```
Comment (mbrodala) [inline Classes/Command/FluidPrecompilerServiceTrait.php:10]: I'm not sure we really need this trait; this could also be done with an abstract command which provides this getter.

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+use Symfony\Component\Console\Logger\ConsoleLogger;
+use Symfony\Component\Console\Output\Output;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use TYPO3Fluid\Fluid\Core\Cache\FluidCacheWarmupResult;
+
+/**
+ * A console command for precompiling Fluid templates
+ */
+class FluidTemplatePrecompileCommand extends Command
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php:16]: Should be named `FluidPrecompileCommand` since it also warms up partials and layouts, see https://github.com/TYPO3/Fluid/blob/f583bad/src/Core/Cache/StandardCacheWarmer.php#L77-L79

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+{
+    use FluidPrecompilerServiceTrait;
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setDescription('Precompile Fluid templates')
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php]: `... templates, partials and layouts`

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+            ->addArgument(
+                'extension',
+                InputArgument::OPTIONAL,
+                'Extension key of which the fluid template should be compiled, leave empty to compile all',
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php]: `... the Fluid resources ...` (especially _Fluid_, not _fluid_. ;-) )

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                'fail',
+                null,
+                InputOption::VALUE_NONE,
+                'Fail with a non-zero return code if any template could not be compiled'
+            )
+            ->addOption(
+                'show-only-failed',
+                null,
+                InputOption::VALUE_NONE,
+                'Show only failed fluid templates from compiling. Only in conjunction with -v'
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php]: `... Fluid templates which failed to compile.`

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            )
+            ->addOption(
+                'limit',
+                'l',
+                InputOption::VALUE_REQUIRED,
+                'Limit of template files of each extension shown. Only in conjunction with -v',
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php]: `... to show.`

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                5
+            )
+            ->addOption(
+                'disable-error-handler',
+                null,
+                InputOption::VALUE_NONE,
+                'Temporary disable any registered error handler for compiling'
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php]: `Temporarily`

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        $onlyFailed = $input->getOption('show-only-failed');
+        $limit = $input->getOption('limit');
+        $disableErrorHandlers = $input->getOption('disable-error-handler');
+
+        // Test exception handlers
+        if (set_error_handler(function() {}) && !$disableErrorHandlers) {
+            $outputHelper->warning(
+                "\nIt seems like there is an error handler registered."
+                . "\nIn some cases this can prevent the precompiling to run properly."
+                . "\nAn ErrorHandler might raise an exception for a PHP Warning."
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php:79]: Here I'd suggest the `--disable-error-handler` flag.

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+            );
+
+            restore_error_handler();
+        }
+
+        $result = $this->getPrecompilerService()->warmup($extension);
+
+        // General information
+        $templatesCount = 0;
+        foreach ($result as $extensionResult) {
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php:89]: This could be done with `array_map()` and `array_sum()`

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            $templatesCount += count($extensionResult);
+        }
+
+        $uncompilable = array_column($result, 'uncompilable');
+        $uncompilable = array_sum($uncompilable);
+
+        $outputHelper->section('General');
+        $outputHelper->text(vsprintf('Compiled %d extension(s) with %d template(s) in total.', [count($result), $templatesCount]));
+        $outputHelper->text(vsprintf('%d template(s) could not be compiled.', [$uncompilable]));
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php:98]: This should be shown only if necessary and as warning/error.

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+                );
+            }
+        }
+
+        if ($disableErrorHandlers) {
+            restore_error_handler();
+        }
+
+        if ($fail && $uncompilable > 0) {
+            throw new \Exception(sprintf('Could not compile %d templates.', $uncompilable), 1518205183);
```
Comment (mbrodala) [inline Classes/Command/FluidTemplatePrecompileCommand.php:143]: Please use `return` instead here.

File: Configuration/Commands.php [hunk 15]
```diff
@@ -16,5 +16,11 @@
     ],
     'referenceindex:update' => [
         'class' => \TYPO3\CMS\Backend\Command\ReferenceIndexUpdateCommand::class
+    ],
+    'fluidprecompile:compile' => [
```
Comment (mbrodala) [inline Configuration/Commands.php:20]: I'd suggest `fluid:precompile`

File: Configuration/Commands.php [hunk 16]
```diff
@@ -16,5 +16,11 @@
...
+        'class' => \NamelessCoder\CmsFluidPrecompilerModule\Command\FluidTemplatePrecompileCommand::class
+    ],
+    'fluidprecompile:profile' => [
```
Comment (mbrodala) [inline Configuration/Commands.php:23]: Here I'd use `fluid:profile`

**Review by mbrodala — CHANGES_REQUESTED** [review]

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 11 above)
Comment (jdoubleu) [reply Classes/Command/FluidTemplatePrecompileCommand.php:79]: It already applies (see the if-condition): If this flag is provided the warning won't be displayed.

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 12 above)
Comment (jdoubleu) [reply Classes/Command/FluidTemplatePrecompileCommand.php:89]: `array_sum` would use the array's values. But these are arrays too (for each extension an array of templates).

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 13 above)
Comment (jdoubleu) [reply Classes/Command/FluidTemplatePrecompileCommand.php:98]: For me this is some general information about the procompiling job.

The command will also fail if the amount of uncompilable templates is greater than 0 and the `--fail` flag was provided.

How about showing this for the first verbosity level? When using `-v` this will be shown.

File: Configuration/Commands.php (hunk 15 above)
Comment (jdoubleu) [reply Configuration/Commands.php:20]: I'm not sure if this is going to conflict with the EXT:fluid.
This extension is name `fluid_precompiler_module` so this should be the commands vendor, shouldn't it?

File: Configuration/Commands.php (hunk 16 above)
Comment (jdoubleu) [reply Configuration/Commands.php:23]: see above.

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 13 above)
Comment (jdoubleu) [reply Classes/Command/FluidTemplatePrecompileCommand.php:98]: I see this as an important information, even when no verbose output is requested and the command should not fail.
You may want to warmup your caches, including the Fluid resources, and have some nice metric.

I'm also against the `-v` option because this also shows the detailed compilation report where you can already extract this metric.

How do you think about this?

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 14 above)
Comment (jdoubleu) [reply Classes/Command/FluidTemplatePrecompileCommand.php:143]: Do you mean to remove the exception and return an invalid status code or to return the exception?
Latter would be wrong because the Command's `execute()` function should only return a status code.

I'd prefer a thrown exception because it is more highlighted in a console. A simple non-zero return code might not be as noticeable as a big red box.

**Comment by jdoubleu** [comment]
Any updates on this?

**Review by mbrodala — DISMISSED** [review]

**Comment by mbrodala** [comment]
@NamelessCoder are your requested changes still valid? From my POV the current state works quite well (really checked on CLI), small details like strict types can be added later.

File: Classes/Command/FluidPrecompilerServiceTrait.php (hunk 4 above)
Comment (NamelessCoder) [reply Classes/Command/FluidPrecompilerServiceTrait.php:10]: Duplicating the getter in two classes is also perfectly acceptable if the alternative is to create a single-method trait or abstract class. In fact, I'd say duplication is preferred for single-line, single-function cases.

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 5 above)
Comment (NamelessCoder) [reply Classes/Command/FluidTemplatePrecompileCommand.php:16]: Agree!

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+
+        // Get arguments and options
+        $extension = $input->getArgument('extension');
+        $fail = $input->getOption('fail');
+        $onlyFailed = $input->getOption('show-only-failed');
+        $limit = $input->getOption('limit');
+        $disableErrorHandlers = $input->getOption('disable-error-handler');
+
+        // Test exception handlers
+        if (set_error_handler(function() {}) && !$disableErrorHandlers) {
```
Comment (NamelessCoder) [inline Classes/Command/FluidTemplatePrecompileCommand.php:75]: This really should not be necessary - the default behavior is to catch exceptions internally and analyse them. We should not need to replace the error handler: if an exception really is raised, it is an error in the templates or an incompatibility which prevents it from being parsed without errors.

FYI: Fluid has an internal error handling device but it is *not* a PHP registered error handler. It is quite selective about which exceptions are actual errors and which are parsing errors - and only "real" errors would break in a way that triggers whichever error handler is registered. Additionally, PHP warnings must definitely not be suppressed because they indicate real errors (such as not checking input types, abusing parse-time execution, etc.)

Therefore I would rather we do not make it possible to manipulate this low level of the runtime. The right way is to show the errors and expect the user to fix them.

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 14 above)
Comment (NamelessCoder) [reply Classes/Command/FluidTemplatePrecompileCommand.php:143]: Failure in this CLI task is by definition supposed to be an error return code, not a runtime exception. The reason behind that is that an exception is indistinguishable from an internal error, whereas a good old return code communicates exactly what it should: that the task failed because you told it to fail if it couldn't compile.

However, a message to go along with that return code certainly wouldn't hurt.

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+        }
+    }
+
+    /**
+     * Formats the given input in a readable format
+     *
+     * @param mixed $data
+     * @return string
+     */
+    protected static function formatReadable($data)
```
Comment (NamelessCoder) [inline Classes/Command/FluidTemplatePrecompileCommand.php:153]: This method does not need to be static - and there is only a single usage. I'd remove it.

File: Configuration/Commands.php (hunk 15 above)
Comment (NamelessCoder) [reply Configuration/Commands.php:20]: Fluid itself will never provide a `precompile` or `profile` command - it's outside the responsibility of the engine package, so there's no risk in using those. If anything, Fluid will at one time provide a `render` command and that's about it.

File: Classes/Command/FluidTemplateProfileCommand.php [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+                    'efficiency' => $helper['efficiency']
+                ];
+            }
+
+            $outputHelper->table(
+                array_keys($rows[0]),
+                $rows
+            );
+        } else {
+            $logger->info('No ViewHelpers used.');
```
Comment (NamelessCoder) [inline Classes/Command/FluidTemplateProfileCommand.php:133]: This should also be put to console via $output

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 12 above)
Comment (NamelessCoder) [reply Classes/Command/FluidTemplatePrecompileCommand.php:89]: What Mathias meant was `$templatesCount = array_sum(array_map('count', $result));` does the same. First it creates an array of counts of results for each extension, then sums those up.

File: Classes/Command/FluidTemplatePrecompileCommand.php (hunk 13 above)
Comment (NamelessCoder) [reply Classes/Command/FluidTemplatePrecompileCommand.php:98]: It's vital information and should be shown. The trick is to only error out if failures are wanted - but the information that you do have uncompilable templates is pretty important, I agree.

File: Classes/Command/FluidTemplatePrecompileCommand.php [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+        $templatesCount = 0;
+        foreach ($result as $extensionResult) {
+            $templatesCount += count($extensionResult);
+        }
+
+        $uncompilable = array_column($result, 'uncompilable');
+        $uncompilable = array_sum($uncompilable);
+
+        $outputHelper->section('General');
+        $outputHelper->text(vsprintf('Compiled %d extension(s) with %d template(s) in total.', [count($result), $templatesCount]));
```
Comment (NamelessCoder) [inline Classes/Command/FluidTemplatePrecompileCommand.php:97]: A more appropriate error message would be `Compiled %d template(s) in %d extension(s)`.

File: Classes/Command/FluidTemplateProfileCommand.php [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+
+        // Get arguments and options
+        $extension = $input->getArgument('extension');
+        $template = $input->getArgument('template');
+
+        $result = $this->getPrecompilerService()->profileTemplateFile($template, $extension);
+
+        // General information
+        $outputHelper->section('General');
+        $rows = [
```
Comment (NamelessCoder) [inline Classes/Command/FluidTemplateProfileCommand.php:54]: This looks almost entirely like `$rows = [$result]` would do the same; except for squashing the sections which could be done afterwards. Since `$result['sections']` doesn't get used below, you could simply do:

```php
$result['sections'] = count($result['sections']);
$outputHelper->table(['Info', 'Value'], [$result]);
```

To replace lines 52-88.

**Review by NamelessCoder — CHANGES_REQUESTED** [review]
Needs a bit more love, see comments.

General CGL-related comments:

* Use `!$var` instead of `false === $var` when it's a boolean. When it may be not-boolean, `$var === false`.
* Avoid redundant comments, for example single words that contain the same text as the only string argument used directly below.
* Remove the double line breaks.
* `sprintf` is preferred over `vsprintf`; there's a much longer explanation but the short version is: `sprintf` when your arguments for the formatted string are explicit and not dynamically built. Use `vsprintf` when your arguments may be dynamic (as in: you don't know the exact argument count and types).
* `PHP_EOL` is always preferred over `"\n"`.
* Only use double-quoted strings when you must; using `PHP_EOL` means you almost never have to double quote.
